src/backend/distributed/metadata/pg_get_object_address_12_13_14.c

```diff
@@ -0,0 +1,536 @@
+		{
+			objnode = (Node *) name;
+			break;
+		}
+
+		case OBJECT_ACCESS_METHOD:
+		case OBJECT_DATABASE:
+		case OBJECT_EVENT_TRIGGER:
+		case OBJECT_EXTENSION:
+		case OBJECT_FDW:
+		case OBJECT_FOREIGN_SERVER:
+		case OBJECT_LANGUAGE:
+		case OBJECT_PUBLICATION:
+		case OBJECT_ROLE:
+		case OBJECT_SCHEMA:
+		case OBJECT_SUBSCRIPTION:
+		case OBJECT_TABLESPACE:
+		{
+			if (list_length(name) != 1)
+			{
+				ereport(ERROR,
+						(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+						 errmsg("name list length must be exactly %d", 1)));
+			}
+			objnode = linitial(name);
+			break;
+		}
+
+		case OBJECT_TYPE:
+		case OBJECT_DOMAIN:
+		{
+			objnode = (Node *) typename;
+			break;
+		}
+
+		case OBJECT_CAST:
+		case OBJECT_DOMCONSTRAINT:
+		case OBJECT_TRANSFORM:
+		{
+			objnode = (Node *) list_make2(typename, linitial(args));
+			break;
+		}
+
+		case OBJECT_PUBLICATION_REL:
+		{
+			objnode = (Node *) list_make2(name, linitial(args));
+			break;
+		}
+
+		case OBJECT_USER_MAPPING:
+		{
+			objnode = (Node *) list_make2(linitial(name), linitial(args));
+			break;
+		}
+
+		case OBJECT_DEFACL:
+		{
+			objnode = (Node *) lcons(linitial(args), name);
+			break;
+		}
+
+		case OBJECT_AMOP:
+		case OBJECT_AMPROC:
+		{
+			objnode = (Node *) list_make2(name, args);
+			break;
+		}
+
+		case OBJECT_FUNCTION:
+		case OBJECT_PROCEDURE:
+		case OBJECT_ROUTINE:
+		case OBJECT_AGGREGATE:
+		case OBJECT_OPERATOR:
+		{
+			ObjectWithArgs *owa = makeNode(ObjectWithArgs);
+
+			owa->objname = name;
+			owa->objargs = args;
+			objnode = (Node *) owa;
+			break;
+		}
+
+		case OBJECT_LARGEOBJECT:
+		{
+			/* already handled above */
+			break;
+		}
+
+			/* no default, to let compiler warn about missing case */
+	}
+
+	if (objnode == NULL)
+	{
+		elog(ERROR, "unrecognized object type: %d", type);
+	}
+
+	ObjectAddress addr = get_object_address(type, objnode,
+											&relation, AccessShareLock, false);
+
+	/* CITUS CODE BEGIN */
+	ErrorIfCurrentUserCanNotDistributeObject(type, &addr, objnode, &relation);
+
+	/* CITUS CODE END */
+
+	/* We don't need the relcache entry, thank you very much */
+	if (relation)
+	{
+		relation_close(relation, AccessShareLock);
+	}
+
+	/* CITUS CODE BEGIN */
+	return addr;
+
+	/* CITUS CODE END */
+}
+
+
+/*
+ * ErrorIfCurrentUserCanNotDistributeObject checks whether current user can
+ * distribute object, if not errors out.
+ */
+static void
+ErrorIfCurrentUserCanNotDistributeObject(ObjectType type, ObjectAddress *addr,
+										 Node *node, Relation *relation)
+{
+	Oid userId = GetUserId();
+	AclMode aclMaskResult = 0;
+	bool passAclCheck = false;
+	Oid idToCheck = InvalidOid;
+
+	/* Since we don't handle sequences like object, add it separately */
+	if (!(SupportedDependencyByCitus(addr) || type == OBJECT_SEQUENCE))
+	{
+		ereport(ERROR, (errmsg("Object type %d can not be distributed by Citus", type)));
+	}
+
+	switch (type)
+	{
+		case OBJECT_ACCESS_METHOD:
+		{
+			if (IsObjectAddressOwnedByExtension(addr, NULL))
```

Comment:
I cannot follow the check for `OBJECT_ACCESS_METHOD`. So, if the object is owned by extension, we throw error, else we don't even check the ACLs.

I guess, the check should have been (!IsObject ...)?

Also, do we only allow access methods that comes with extensions? I mean, do we propagate `CREATE ACCESS METHOD`  or not?